// igniteui-webcomponents/components/select/select-group.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, html } from 'lit';
import { property, queryAssignedElements } from 'lit/decorators.js';
import { themes } from '../../theming/theming-decorator.js';
import { watch } from '../common/decorators/watch.js';
import { registerComponent } from '../common/definitions/register.js';
import { styles } from '../dropdown/themes/dropdown-group.base.css.js';
import { all } from '../dropdown/themes/group.js';
let IgcSelectGroupComponent = class IgcSelectGroupComponent extends LitElement {
    static register() {
        registerComponent(this);
    }
    constructor() {
        super();
        this.disabled = false;
        this._internals = this.attachInternals();
        this._internals.role = 'group';
        this.observer = new MutationObserver(this.updateControlledItems.bind(this));
    }
    disconnectedCallback() {
        this.observer.disconnect();
        super.disconnectedCallback();
    }
    async firstUpdated() {
        await this.updateComplete;
        this.controlledItems = this.activeItems;
        this.items.forEach((i) => {
            this.observer.observe(i, {
                attributes: true,
                attributeFilter: ['disabled'],
            });
        });
        this.disabledChange();
    }
    updateControlledItems(mutations) {
        mutations.forEach((mutation) => {
            const item = mutation.target;
            if (!this.disabled) {
                this.controlledItems = this.activeItems;
            }
            if (this.disabled && !item.disabled) {
                item.disabled = true;
            }
        });
    }
    disabledChange() {
        this._internals.ariaDisabled = `${this.disabled}`;
        for (const item of this.controlledItems) {
            item.disabled = this.disabled;
        }
    }
    render() {
        return html `
      <label part="label">
        <slot name="label"></slot>
      </label>
      <slot></slot>
    `;
    }
};
IgcSelectGroupComponent.tagName = 'igc-select-group';
IgcSelectGroupComponent.styles = styles;
__decorate([
    queryAssignedElements({ flatten: true, selector: 'igc-select-item' })
], IgcSelectGroupComponent.prototype, "items", void 0);
__decorate([
    queryAssignedElements({
        flatten: true,
        selector: 'igc-select-item:not([disabled])',
    })
], IgcSelectGroupComponent.prototype, "activeItems", void 0);
__decorate([
    property({ reflect: true, type: Boolean })
], IgcSelectGroupComponent.prototype, "disabled", void 0);
__decorate([
    watch('disabled', { waitUntilFirstUpdate: true })
], IgcSelectGroupComponent.prototype, "disabledChange", null);
IgcSelectGroupComponent = __decorate([
    themes(all)
], IgcSelectGroupComponent);
export default IgcSelectGroupComponent;
