// igniteui-webcomponents/components/common/mixins/option.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { LitElement, html } from 'lit';
import { property, queryAssignedNodes } from 'lit/decorators.js';
import { watch } from '../decorators/watch.js';
export class IgcBaseOptionLikeComponent extends LitElement {
    get _contentSlotText() {
        return this._content.map((node) => node.textContent).join('');
    }
    get value() {
        return this._value ? this._value : this._contentSlotText;
    }
    set value(value) {
        const old = this._value;
        this._value = value;
        this.requestUpdate('value', old);
    }
    disabledChange() {
        this._internals.ariaDisabled = `${this.disabled}`;
    }
    selectedChange() {
        this._internals.ariaSelected = `${this.selected}`;
        this.active = this.selected;
    }
    constructor() {
        super();
        this.active = false;
        this.disabled = false;
        this.selected = false;
        this._internals = this.attachInternals();
        this._internals.role = 'option';
    }
    connectedCallback() {
        super.connectedCallback();
        this.role = 'option';
    }
    render() {
        return html `
      <section part="prefix">
        <slot name="prefix"></slot>
      </section>
      <section part="content">
        <slot></slot>
      </section>
      <section part="suffix">
        <slot name="suffix"></slot>
      </section>
    `;
    }
}
__decorate([
    queryAssignedNodes({ flatten: true })
], IgcBaseOptionLikeComponent.prototype, "_content", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcBaseOptionLikeComponent.prototype, "active", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcBaseOptionLikeComponent.prototype, "disabled", void 0);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcBaseOptionLikeComponent.prototype, "selected", void 0);
__decorate([
    property()
], IgcBaseOptionLikeComponent.prototype, "value", null);
__decorate([
    watch('disabled')
], IgcBaseOptionLikeComponent.prototype, "disabledChange", null);
__decorate([
    watch('selected')
], IgcBaseOptionLikeComponent.prototype, "selectedChange", null);
