// @daelmaak/ngx-gallery/esm2022/lib/directives/media.directive.mjs
import { Directive, EventEmitter, Output, } from '@angular/core';
import * as i0 from "@angular/core";
export class MediaDirective {
    constructor(hostRef) {
        this.mediaLoadError = new EventEmitter();
        this.onLoad = (ev) => {
            if (ev.type === 'error') {
                this.mediaLoadError.emit();
            }
        };
        this.nativeEl = hostRef.nativeElement;
    }
    ngOnInit() {
        this.nativeEl.addEventListener('error', this.onLoad, true);
    }
    ngOnDestroy() {
        this.nativeEl.removeEventListener('error', this.onLoad, true);
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "16.1.8", ngImport: i0, type: MediaDirective, deps: [{ token: i0.ElementRef }], target: i0.ɵɵFactoryTarget.Directive }); }
    static { this.ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "16.1.8", type: MediaDirective, isStandalone: true, selector: "[media]", outputs: { mediaLoadError: "mediaLoadError" }, ngImport: i0 }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "16.1.8", ngImport: i0, type: MediaDirective, decorators: [{
            type: Directive,
            args: [{
                    selector: '[media]',
                    standalone: true,
                }]
        }], ctorParameters: function () { return [{ type: i0.ElementRef }]; }, propDecorators: { mediaLoadError: [{
                type: Output
            }] } });
